armi/interfaces.py

```diff
@@ -77,6 +81,143 @@ class STACK_ORDER:  # pylint: disable=invalid-name, too-few-public-methods
+
+        Parameters
+        ----------
+        val : _SUPPORTED_TYPES
+            the most recent value for computing convergence critera. Is commonly equal to interface.getTightCouplingValue()
+
+        Returns
+        -------
+        boolean
+            True (False) interface is (not) converged
+
+        Notes
+        -----
+        - On convergence, this class is automatically reset to its initial condition to avoid retaining
+          or holding a stale state. Calling this method will increment a counter that when exceeded will
+          clear the state. A warning will be reported if the state is cleared prior to the convergence
+          criteria being met.
+        - For computing convergence of arrays, only up to 2D is allowed. 3D arrays would arise from considering
+          component level parameters. However, converging on component level parameters is not supported at this time.
+
+        Raises
+        ------
+        ValueError
+            If the previous iteration value has not been assigned. The ``storePreviousIterationValue`` method
+            must be called first.
+        RuntimeError
+            Only support calculating norms for up to 2D arrays.
+        """
+        if self._previousIterationValue is None:
+            raise ValueError(
+                f"Cannot check convergence of {self} with no previous iteration value set. "
+                f"Set using `storePreviousIterationValue` first."
+            )
+
+        previous = self._previousIterationValue
+
+        # calculate convergence of val and previous
+        if isinstance(val, (int, float)):
+            self.eps = abs(val - previous)
+        else:
+            dim = self._getListDimension(val)
+            if dim == 1:  # 1D array
+                self.eps = norm(numpy.subtract(val, previous), ord=2)
+            elif dim == 2:  # 2D array
+                epsVec = []
+                for old, new in zip(previous, val):
+                    epsVec.append(norm(numpy.subtract(old, new), ord=2))
+                self.eps = norm(epsVec, ord=numpy.inf)
+            else:
+                raise RuntimeError(
+                    "Currently only support up to 2D arrays for calculating convergence of arrays."
+                )
+
+        # Check if convergence is satisfied. If so, or if reached max number of iters, then
+        # reset the number of iterations
+        converged = self.eps < self.tolerance
+        if converged:
+            self._numIters = 0
+        else:
+            self._numIters += 1
+            if self._numIters == self.maxIters:
+                runLog.warning(
+                    f"Maximum number of iterations for {self.parameter} reached without convergence!"
+                    f"Prescribed convergence criteria is {self.tolerance}."
+                )
+                self._numIters = 0
+
+        return converged
+
+    @staticmethod
+    def _getListDimension(listToCheck: list, dim: int = 1):
```

Comment:
Oh, fun fact.

You can't have a "static method" that is also "private".  It would never get used. 

I see when you call this method, you have to do a `# pylint: disable=protected-access`.  Based on that, I recommend removing the leading underscore. All static methods are public.

Reply:
Oh! That explains that. I was very confused by that Pylint message...

Reply:
https://github.com/terrapower/armi/pull/1033/commits/ae34fbfddcf45a175e4cafc1dc4189823c25c80b